drivers: wifi: esp_at: support listening UDP sockets in active mode

mniestroj opened the pull request:
So far receving was possible in active mode, but IP and port information
was not fetched nor propagated to application layer. Support fetching that
information even in active mode, so that using `bind()` with `recvfrom()` is
working accordingly with UDP server listen flow.

mniestroj committed:
drivers: wifi: esp_at: introduce esp_pull_long() helper

Use it to get information about RSSI. In subsequent commits this helper
will be used to get other information as well.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: wifi: esp_at: handle ':' in esp_pull()

Treat ':' similar as ',', '\r' and '\n', so that AT response parameter
parsing can be gracefully stopped.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: wifi: esp_at: utilize esp_pull*() in cmd_ipd_parse_hdr()

Use common utilities for consuming parameters from AT response string.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: wifi: esp_at: move socket allocation to cmd_ipd_parse_hdr()

This is just a step forward to make cmd_ipd_parse_hdr() capable of parsing
IP address and port when CONFIG_WIFI_ESP_AT_PASSIVE_MODE=n. After obtaining
this information (IP and port) it will be simply much easier to write this
information into 'struct esp_socket' directly, instead of passing to upper
layer.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mniestroj committed:
drivers: wifi: esp_at: support listening UDP sockets in active mode

So far receving was possible in active mode, but IP and port information
was not fetched nor propagated to application layer. Support fetching that
information even in active mode, so that using bind() with recvfrom() is
working accordingly.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

jukkar commented on drivers/wifi/esp_at/esp.c:
```diff
@@ -816,9 +816,44 @@ static int cmd_ipd_parse_hdr(struct esp_data *dev,
+		}
+
+		recv_addr->sin_family = AF_INET;
+		recv_addr->sin_port = ntohs(port);
```
The sockaddr_in struct expects the port number to be in network byte order, so the `ntohs()` could be omitted.
See https://www.man7.org/linux/man-pages/man3/sockaddr.3type.html for details.

mniestroj replied:
Good catch! I have basically reused part of the code from https://github.com/zephyrproject-rtos/zephyr/blob/48d69b4e7e0add3b00320d4568149213ff709867/drivers/wifi/esp_at/esp_offload.c#L547 which handles "passive mode" in the driver.

In both cases `port` is in host (native) byte order, after parsing a string to integer. This means that `htons()` should be used intead of `ntohs()` here. Everything worked so far, since both macros (`htons` and `ntohs`) are actually `BWAP_16()` macro invocations.

I have changed both invocations of `ntohs()` to be `htons()` instead.

mniestroj committed:
drivers: wifi: esp_at: s/ntohs/htons/ after parsing port number

'struct sockaddr_in' should contain port number in network byte order. This
means that htons() macro should be used, instead of ntohs().

This was working before since both htons() and ntohs() end up in calling
BSWAP_16().

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

jukkar reviewed (approved):
LGTM
